pkg/controller/storagenode/storagenode_test.go

```diff
@@ -0,0 +1,615 @@
+		CustomResourceDefinitions().
+		List(metav1.ListOptions{})
+	require.NoError(t, err)
+	require.Len(t, crds.Items, 1)
+	for _, crd := range crds.Items {
+		require.NotEqual(t, nodeStatusCRDName, crd.Name)
+	}
+}
+
+func TestReconcile(t *testing.T) {
+	logrus.SetLevel(logrus.TraceLevel)
+	mockCtrl := gomock.NewController(t)
+	defaultQuantity, _ := resource.ParseQuantity("0")
+	testNS := "test-ns"
+	clusterName := "test-cluster"
+	cluster := &corev1alpha1.StorageCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      clusterName,
+			Namespace: testNS,
+		},
+	}
+
+	controllerKind := corev1alpha1.SchemeGroupVersion.WithKind("StorageCluster")
+	clusterRef := metav1.NewControllerRef(cluster, controllerKind)
+	testStorageNode := "node1"
+	testStoragelessNode := "node2"
+	storageNode := &corev1alpha1.StorageNode{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            testStorageNode,
+			Namespace:       cluster.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*clusterRef},
+		},
+		Status: corev1alpha1.NodeStatus{
+			Phase: string(corev1alpha1.NodeInitStatus),
+			Storage: corev1alpha1.StorageStatus{
+				TotalSize: *resource.NewQuantity(20971520, resource.BinarySI),
+				UsedSize:  *resource.NewQuantity(10971520, resource.BinarySI),
+			},
+			Conditions: []corev1alpha1.NodeCondition{
+				{
+					Type:   corev1alpha1.NodeStateCondition,
+					Status: corev1alpha1.NodeOnlineStatus,
+				},
+			},
+		},
+	}
+
+	storageLessNode := &corev1alpha1.StorageNode{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            testStoragelessNode,
+			Namespace:       cluster.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*clusterRef},
+		},
+		Status: corev1alpha1.NodeStatus{
+			Phase: string(corev1alpha1.NodeInitStatus),
+			Storage: corev1alpha1.StorageStatus{
+				TotalSize: defaultQuantity,
+				UsedSize:  defaultQuantity,
+			},
+			Conditions: []corev1alpha1.NodeCondition{
+				{
+					Type:   corev1alpha1.NodeStateCondition,
+					Status: corev1alpha1.NodeOnlineStatus,
+				},
+			},
+		},
+	}
+
+	k8sClient := testutil.FakeK8sClient(storageNode, storageLessNode, cluster)
+	recorder := record.NewFakeRecorder(10)
+	driver := testutil.MockDriver(mockCtrl)
+	driver.EXPECT().GetSelectorLabels().Return(nil).AnyTimes()
+	driver.EXPECT().String().Return("ut-driver").AnyTimes()
+
+	// reconcile storage node
+	podNode1 := createStoragePod(cluster, "pod-node1", testStorageNode, nil, clusterRef)
+	k8sClient.Create(context.TODO(), podNode1)
+
+	podNode2 := createStoragePod(cluster, "pod-node2", testStoragelessNode, nil, clusterRef)
+	k8sClient.Create(context.TODO(), podNode2)
+
+	controller := Controller{
+		client:   k8sClient,
+		recorder: recorder,
+		Driver:   driver,
+	}
+
+	request := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      testStorageNode,
+			Namespace: testNS,
+		},
+	}
+	result, err := controller.Reconcile(request)
+	require.NoError(t, err)
+	require.NotNil(t, result)
```

Comment:
nit: result is not a pointer so it will always be non-nil even in case of error. If you want, you can check that it is empty.